cleverhans/attacks_tf.py

```diff
@@ -1918,3 +1918,69 @@ def margin_logit_loss(model_logits, label, num_classes=10):
         logits_with_target_label_neg_inf, axis=-1)
     loss = highest_nonlabel_logits - label_logits
     return loss
+
+
+def _apply_transformation(x, dx, dy, angle, batch_size):
+    # Map a transformation onto the input
+    angle *= np.pi / 180
+    height = x.get_shape().as_list()[1]
+    width = x.get_shape().as_list()[2]
+    M = np.array([1, 0, -dx*height,
+                  0, 1, -dy*width, 0, 0] * batch_size, dtype=np.float32)
+    theta = tf.constant(M, shape=(batch_size, 8))
+
+
+    # Pad the image to prevent two-step rotation / translation
+    x = tf.pad(x, [[0, 0], [50, 50], [50, 50], [0, 0]], 'CONSTANT')
```

Comment:
It could make sense to have the padding depend on the image size. If the image is large, we could still end up with chopped corners.